### PR TITLE
[RL] Capture GDN prefill with stable replay metadata

### PR DESCRIPTION
**Pull request by drisspg**
**Replaced by #4738**, with the same implementation on an upstream `pytorch/torchtitan` branch. The active native stack is #4736 -> #4737 -> #4738 (stack #4739). This PR is retained for its prior review history.

---

### Agent notes

Capture GDN prefill, mixed batches, and specialized decode through **vLLM's native FULL graph lifecycle**. Removing the eager decorator alone is unsafe: captured Python metadata lookups do not run again during replay.

**Numerics: exact matches, no tolerance relaxation.** On Qwen3.5-0.8B TP1/DP1, eager versus FULL replay matched all **72 generated tokens**, returned logprobs, and **936 per-step/per-layer used-slot conv/SSM state hashes**. Separately, **14 regular/batch-invariant GDN cases** matched the pre-redesign eager implementation bitwise for outputs and entire conv/SSM cache storage, including padded inputs. These are the tested cases, not a TP4 correctness claim.

- A layer-local Attention Gym metadata builder reuses upstream `GDNAttentionMetadata` directly and implements `build()` and `build_for_cudagraph_capture()`, staging request offsets, cache slots, and freshness into persistent buffers. Reuse native common metadata and cache allocation; do not override the global GDN backend.
- Remove the custom replay wrapper, forward-context rewriting, legacy metadata path, and graph-mode-dependent backend selection. Native nested FULL/PIECEWISE wrappers, graph pools, profiling, cache recreation, capture, and replay remain intact.
- Keep one GDN metadata/kernel-selection contract for eager and captured execution. Preserve fused pure-decode kernels and batch-invariant recurrent decode rather than substituting the numerically different packed prefill path. PIECEWISE still honors native eager attention boundaries; select FULL for full-forward capture.
- Define padding explicitly: packed batches get a null-tail request, decode gets one offset interval per physical row, and non-FULL computation stays within the prepared token extent.
- Reuse Attention Gym `gate_transform(kind="softplus", impl="reference")`, preserving the existing FP32 formula. The compatible Gym 0.0.9 release is supplied by #4736. `bound_gate`'s bounded sigmoid is not equivalent.

### Narrow upstream gap

At vLLM `c6fa1f05d`, configured FULL discards the uniform-decode discriminator when initializing/selecting graph keys. Equal token-count prefill and specialized decode would collide. The native graph wrapper already supports distinct keys.

A small TorchTitan dispatcher adapter therefore delegates decode key generation/selection to a native `FULL_DECODE_ONLY` dispatcher and registers those descriptors for native capture. No capture/replay algorithm is copied. Remove this adapter when upstream exposes FULL decode specialization.

## Before / after prefill trace

**[Merged, time-aligned Perfetto trace (internal)](https://fburl.com/perfetto/qhb6334q)** — labeled BEFORE / AFTER; upload has a 28-day TTL.

Matched **Qwen3.5-0.8B, TP1/DP1, BF16 on one GB200**: batch 32, input 4096, output 1, scheduler budget 2048 (**64 prefill chunks**), max model length 8192, compile/prefix caching/async scheduling off. Actual source snapshots: pre-PR `c48650f06` with FULL_AND_PIECEWISE versus `1ac48cbbf` with FULL, using the same weights and dependencies.

| Metric | Before | After |
|---|---:|---:|
| Unprofiled batch latency, median of 10 runs | 1.869 s | 0.809 s |
| CUDA graph launches per batch | 1,600 | 64 |
| Direct kernel launches per batch | 30,208 | 2,048 |
| Eager breaks per chunk | 24 | 0 |

**2.31x speedup / 56.7% lower latency** for this workload. Timing excludes profiling, initialization/capture, and two complete workload warmups; native graph warmups remain 1. Generated IDs, text, and finish reasons matched across all 14 batches per mode. Launch counts come from separate steady-state traces; break counts from an untimed runtime audit.

This is one sequential process pair on the same exclusively held GPU with normal DVFS, **not the original W2 27B/TP4 reproduction**. FULL captures both 18 GDN and 6 standard-attention layers, so the improvement is not attributable solely to GDN. Profiling distorts the two modes unequally: use the unprofiled timings above, not trace durations, for speedup.

## Validation

**5 tests passed with published Attention Gym 0.0.9:** four CPU guards plus one real-engine eager/FULL test.

```sh
TORCHTITAN_QWEN3_5_0_8B_HF_PATH=/path/to/Qwen3.5-0.8B \
.venv/bin/pytest -v -ra \
  torchtitan/experiments/rl/tests/test_gdn_graph_metadata.py \
  torchtitan/experiments/rl/tests/test_gdn_cudagraph.py
```

The engine test uses Qwen3.5-0.8B, GB200, TP1/DP1, BF16, compile/prefix caching off, and one capture warmup. It exercises real model loading, graph-memory profiling, cache recreation, capture, and execution:

- Exact equality for **72 generated tokens**, returned logprobs, and **936 used-slot conv/SSM state hashes**.
- **25 actual FULL replays** across prefill, mixed batches, and padded uniform decode; changed cache slots and an over-capacity NONE forward.
- Asserts the native outer FULL/inner PIECEWISE wrappers and actual graph replay, not just dispatch labels.

Additional one-off GPU validation compared the redesigned path against the pre-redesign eager implementation: 10 regular/batch-invariant decode, prefill, and mixed cases plus 4 padded cases passed bitwise output and full conv/SSM cache comparisons. The padded cases use NaN-poisoned input tails and check zero output padding. These larger diagnostic scripts are not added to the repository. The before/after performance runs separately checked exact generated IDs, text, and finish reasons across all 14 batches per mode; they did not collect logits or state hashes inside timed/profiled regions. The numerical checks above are independent of the timing measurements.

Staged-file pre-commit checks passed with the global mutating Pyrefly hook skipped; separate targeted Pyrefly passed for all three production files (0 errors). Broad local checks retain unrelated checkpointing dependency/type errors; no unrelated files were changed.

### Tested dependencies

- Torch `2.15.0.dev20260913+cu130`; torchvision `0.30.0.dev20260914+cu130`.
- vLLM distribution `1.0.0.dev20260913+cu130` (packaged runtime `0.1.dev1+g9f03b510c.d20260915`); inspected runner/dispatcher/metadata sources match the cited `c6fa1f05d` revision.
- Attention Gym **0.0.9** from PyPI; FA4 commit `383cbcba`. The shared performance traces retain their explicitly recorded earlier source/dependency provenance; correctness was rerun against the published release.
- TVM FFI `0.1.13.post3`, with an explicit override of vLLM's older pin and exclusion of unused incompatible TileLang/tokenspeed packages. Setup commands are in the existing RL README; no installed-vLLM source patches.

## Scope / remaining validation

Broader FULL-prefill capability is enabled for DP1 without DBO/ubatching; other configurations retain decode-only capability and native graph-mode resolution. Speculative decoding and intermediate (`mamba_cache_mode="all"`) prefix-cache checkpoints fail explicitly. At least one FULL capture warmup is required.

The new tuple is validated at TP1 only. TP2/TP4, SP/DP/DBO, cache `align`, other attention backends, and graph-memory estimate accuracy across both routine families remain unvalidated. Keep this PR draft pending target-topology validation. The standalone graph guide remains removed.

### CODE REVIEW

File: .ci/docker/requirements.txt [hunk 1]
```diff
@@ -4,11 +4,10 @@ tensorboard
 wandb
 tyro >= 1.0.5
 tokenizers >= 0.15.0
-renderers == 0.1.11
 safetensors
 einops
 pillow
 spmd_types==0.2.5
-attn-gym[linear]==0.0.8
+attn-gym[linear] @ git+https://github.com/meta-pytorch/attention-gym.git@1f361d9bb16f96dc37692a33808a4639aff44a20
```
Comment (drisspg) [inline .ci/docker/requirements.txt]: bump before land
